File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/ProfitService.java
package cz.fb.manaus.reactor.profit;

import com.google.common.base.Preconditions;
import cz.fb.manaus.core.category.BetCoverage;
import cz.fb.manaus.core.category.CategoryService;
import cz.fb.manaus.core.model.ProfitRecord;
import cz.fb.manaus.core.model.SettledBet;
import cz.fb.manaus.core.model.Side;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.OptionalDouble;
import java.util.stream.Collectors;

import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toList;


@Service
public class ProfitService {

    @Autowired
    private CategoryService categoryService;
    @Autowired
    private ProfitPlugin profitPlugin;

    public List<ProfitRecord> getProfitRecords(List<SettledBet> bets, Optional<String> projection,
                                               boolean simulationAwareOnly, double chargeRate) {
        var coverage = BetCoverage.from(bets);
        var charges = profitPlugin.getCharges(bets, chargeRate);

        if (projection.isPresent()) {
            bets = categoryService.filterBets(bets, projection.get(), coverage);
        }

        var betRecords = computeProfitRecords(bets, simulationAwareOnly, charges, coverage);
        return mergeProfitRecords(betRecords);
    }

    public List<ProfitRecord> mergeProfitRecords(Collection<ProfitRecord> records) {
        var categories = records.stream()
                .collect(Collectors.groupingBy(ProfitRecord::getCategory));
        return categories.entrySet().stream()
                .map(e -> mergeCategory(e.getKey(), e.getValue()))
                .sorted(Comparator.comparing(ProfitRecord::getCategory))
                .collect(toList());
    }

    public ProfitRecord mergeCategory(String category, Collection<ProfitRecord> records) {
        Preconditions.checkArgument(records.stream().map(ProfitRecord::getCategory)
                .allMatch(category::equals));
        var avgPrice = records.stream()
                .mapToDouble(ProfitRecord::getAvgPrice)
                .average().getAsDouble();
        var theoreticalProfit = records.stream().mapToDouble(ProfitRecord::getTheoreticalProfit).sum();
        var charge = records.stream().mapToDouble(ProfitRecord::getCharge).sum();
        var layCount = records.stream().mapToInt(ProfitRecord::getLayCount).sum();
        var backCount = records.stream().mapToInt(ProfitRecord::getBackCount).sum();
        var coverCount = records.stream().mapToInt(ProfitRecord::getCoverCount).sum();
        var result = new ProfitRecord(category, theoreticalProfit, layCount, backCount, avgPrice, charge);
        if (coverCount > 0) {
            OptionalDouble diff = records.stream().filter(profitRecord -> profitRecord.getCoverDiff() != null)
                    .mapToDouble(ProfitRecord::getCoverDiff).average();
            result.setCoverDiff(diff.getAsDouble());
            result.setCoverCount(coverCount);
        }
        return result;
    }

    private List<ProfitRecord> computeProfitRecords(List<SettledBet> bets, boolean simulationAwareOnly,
                                                    Map<String, Double> charges, BetCoverage coverage) {
        return bets.parallelStream().flatMap(bet -> {
            var categories = categoryService.getSettledBetCategories(bet, simulationAwareOnly, coverage);
            return categories.stream().map(category -> {
                double charge = charges.get(bet.getBetAction().getBetId());
                Preconditions.checkState(charge >= 0, charge);
                return toProfitRecord(bet, category, charge, coverage);
            });
        }).collect(Collectors.toList());
    }

    public ProfitRecord toProfitRecord(SettledBet bet, String category, double chargeContribution, BetCoverage coverage) {
        var type = requireNonNull(bet.getPrice().getSide());
        var price = bet.getPrice().getPrice();
        ProfitRecord result;
        if (type == Side.BACK) {
            result = new ProfitRecord(category, bet.getProfitAndLoss(), 0, 1, price, chargeContribution);
        } else {
            result = new ProfitRecord(category, bet.getProfitAndLoss(), 1, 0, price, chargeContribution);
        }
        var marketId = bet.getBetAction().getMarket().getId();
        var selectionId = bet.getSelectionId();
        if (coverage.isCovered(marketId, selectionId)) {
            var backPrice = coverage.getPrice(marketId, selectionId, Side.BACK);
            var layPrice = coverage.getPrice(marketId, selectionId, Side.LAY);
            result.setCoverDiff(backPrice - layPrice);
            result.setCoverCount(1);
        }
        return result;
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/profit/ProfitService.kt
package cz.fb.manaus.reactor.profit

import cz.fb.manaus.core.category.BetCoverage
import cz.fb.manaus.core.category.CategoryService
import cz.fb.manaus.core.model.ProfitRecord
import cz.fb.manaus.core.model.RealizedBet
import cz.fb.manaus.core.model.Side
import org.springframework.stereotype.Service


@Service
class ProfitService(private val categoryService: CategoryService) {

    fun getProfitRecords(
            bets: List<RealizedBet>,
            projection: String? = null,
            simulationAwareOnly: Boolean
    ): List<ProfitRecord> {
        val coverage = BetCoverage.from(bets)
        val filtered = if (projection == null) bets else categoryService.filterBets(bets, projection)

        val betRecords = computeProfitRecords(filtered, simulationAwareOnly, coverage)
        return mergeProfitRecords(betRecords)
    }

    private fun mergeProfitRecords(records: Collection<ProfitRecord>): List<ProfitRecord> {
        val categories = records.groupBy { it.category }
        return categories.entries
                .map { mergeCategory(it.key, it.value) }
                .sortedBy { it.category }
    }

    fun mergeCategory(category: String, records: Collection<ProfitRecord>): ProfitRecord {
        require(records.map { it.category }
            .all { category == it })
        val avgPrice = records
            .map { it.avgPrice }
            .average()
        val theoreticalProfit = records.sumOf { it.theoreticalProfit }
        val charge = records.sumOf { it.charge }
        val layCount = records.sumOf { it.layCount }
        val backCount = records.sumOf { it.backCount }
        val coverCount = records.sumOf { it.coverCount }
        val result = ProfitRecord(category, theoreticalProfit, avgPrice, charge, layCount, backCount)
        if (coverCount > 0) {
            val diff = records.filter { it.coverIndex != null }.mapNotNull { it.coverIndex }.average()
            result.coverIndex = diff
            result.coverCount = coverCount
        }
        return result
    }

    private fun computeProfitRecords(
            bets: List<RealizedBet>,
            simulationAwareOnly: Boolean,
            coverage: BetCoverage
    ): List<ProfitRecord> {
        return bets.flatMap { bet ->
            val categories = categoryService.getRealizedBetCategories(bet, simulationAwareOnly)
            categories.map {
                val charge = bet.settledBet.commission ?: 0.0
                check(charge >= 0) { charge }
                toProfitRecord(bet, it, charge, coverage)
            }
        }
    }

    fun toProfitRecord(
            bet: RealizedBet,
            category: String,
            chargeContribution: Double,
            coverage: BetCoverage
    ): ProfitRecord {
        val side = bet.settledBet.price.side
        val price = bet.settledBet.price.price
        val result = if (side == Side.BACK) {
            ProfitRecord(category, bet.settledBet.profitAndLoss, price, chargeContribution, 0, 1)
        } else {
            ProfitRecord(category, bet.settledBet.profitAndLoss, price, chargeContribution, 1, 0)
        }
        val marketId = bet.market.id
        val selectionId = bet.settledBet.selectionId
        if (coverage.isCovered(marketId, selectionId)) {
            val backPrice = coverage.getPrice(marketId, selectionId, Side.BACK)
            val layPrice = coverage.getPrice(marketId, selectionId, Side.LAY)
            result.coverIndex = 100 * backPrice / layPrice - 100
            result.coverCount = 1
        }
        return result
    }

}
